tests/python/pants_test/util/test_objects.py

```diff
@@ -106,23 +105,70 @@ def test_none(self):
     with self.assertRaises(ValueError):
       SubclassesOf()
 
+  def test_str_and_repr(self):
+    subclasses_of_b = SubclassesOf(self.B)
+    self.assertEqual("+B", str(subclasses_of_b))
+    self.assertEqual("SubclassesOf(B)", repr(subclasses_of_b))
+
+    subclasses_of_multiple = SubclassesOf(self.A, self.B)
+    self.assertEqual("+(A, B)", str(subclasses_of_multiple))
+    self.assertEqual("SubclassesOf(A, B)", repr(subclasses_of_multiple))
+
   def test_single(self):
     subclasses_of_b = SubclassesOf(self.B)
-    self.assertEqual((self.B,), subclasses_of_b.types)
     self.assertFalse(subclasses_of_b.satisfied_by(self.A()))
     self.assertTrue(subclasses_of_b.satisfied_by(self.B()))
     self.assertFalse(subclasses_of_b.satisfied_by(self.BPrime()))
     self.assertTrue(subclasses_of_b.satisfied_by(self.C()))
 
   def test_multiple(self):
     subclasses_of_b_or_c = SubclassesOf(self.B, self.C)
-    self.assertEqual((self.B, self.C), subclasses_of_b_or_c.types)
     self.assertTrue(subclasses_of_b_or_c.satisfied_by(self.B()))
     self.assertTrue(subclasses_of_b_or_c.satisfied_by(self.C()))
     self.assertFalse(subclasses_of_b_or_c.satisfied_by(self.BPrime()))
     self.assertFalse(subclasses_of_b_or_c.satisfied_by(self.A()))
 
 
+class TypedCollectionTest(TypeConstraintTestBase):
+  def test_str_and_repr(self):
+    collection_of_exactly_b = TypedCollection(Exactly(self.B))
+    self.assertEqual("[=]B", str(collection_of_exactly_b))
+    self.assertEqual("TypedCollection(Exactly(B))",
+                     repr(collection_of_exactly_b))
+
+    collection_of_multiple_subclasses = TypedCollection(
+      SubclassesOf(self.A, self.B))
+    self.assertEqual("[+](A, B)", str(collection_of_multiple_subclasses))
+    self.assertEqual("TypedCollection(SubclassesOf(A, B))",
+                     repr(collection_of_multiple_subclasses))
+
+  def test_collection_single(self):
+    collection_constraint = TypedCollection(Exactly(self.A))
+    self.assertTrue(collection_constraint.satisfied_by([self.A()]))
+    self.assertFalse(collection_constraint.satisfied_by([self.A(), self.B()]))
+    self.assertTrue(collection_constraint.satisfied_by([self.A(), self.A()]))
+
+  def test_collection_multiple(self):
+    collection_constraint = TypedCollection(SubclassesOf(self.B, self.BPrime))
+    self.assertTrue(collection_constraint.satisfied_by([self.B(), self.C(), self.BPrime()]))
+    self.assertFalse(collection_constraint.satisfied_by([self.B(), self.A()]))
+
+    def collection_generator():
+      yield self.A()
+      yield self.C()
+      yield self.A()
+
+    collection_constraint = TypedCollection(Exactly(self.A, self.C))
+    self.assertEquals(
+      (self.A(), self.C(), self.A()),
+      collection_constraint.validate_satisfied_by(tuple(collection_generator())))
```

Comment:
I'm not sure what value the generator adds here?

Reply:
It added value with the `wrapper_type` example to show that it was converted into a tuple. It can probably be removed entirely.

Reply:
Removed!